src/tosyntax/FStar.ToSyntax.Resugar.fs

```diff
@@ -0,0 +1,614 @@
+              resugar_sequence term
+          | Primop (* doesn't seem to be generated by desugar *)
+          | Masked_effect (* doesn't seem to be generated by desugar *)
+          | Meta_smt_pat -> (* nothing special, just resugar the term *)
+              resugar_term e
+          | Mutable_alloc ->
+              let term = resugar_term e in
+              begin match term.tm with
+                | A.Let(A.NoLetQualifier,l,t) -> mk (A.Let(A.Mutable, l, t))
+                | _ -> failwith "mutable_alloc should have let term with no qualifier"
+              end
+          | Mutable_rval ->
+              let fv = S.lid_as_fv C.sread_lid Delta_constant None in
+              begin match (SS.compress e).n with 
+                | Tm_app({n=Tm_fvar fv}, [(term, _)])-> resugar_term term
+                | _ -> failwith "mutable_rval should have app term"
+              end
+      in
+      begin match m with
+      | Meta_pattern pats ->
+          failwith "This case is impossible, if Forall and Exist is properly resugared"
+      | Meta_labeled (l, _, p) ->
+          mk (A.Labeled(resugar_term e, l, p))
+      | Meta_desugared i -> 
+          resugar_meta_desugared i
+      | Meta_named _ 
+      | Meta_monadic _ 
+      | Meta_monadic_lift _ -> 
+          failwith "cases not generated by desugar"
+      end
+    
+    | Tm_unknown _ -> mk A.Wild
+
+and resugar_comp (c:S.comp) : A.term =
+  let mk (a:A.term') : A.term =
+        //augment `a` with its source position
+        //and an Unknown level (the level is unimportant ... we should maybe remove it altogether)
+        A.mk_term a c.pos A.Un
+  in
+  match (c.n) with 
+  | Total (typ, _) ->
+    let t = resugar_term typ in
+    mk (A.Construct(C.effect_Tot_lid, [(t, A.Nothing)]))
+
+  | GTotal (typ, _) ->
+    let t = resugar_term typ in
+    mk (A.Construct(C.effect_GTot_lid, [(t, A.Nothing)]))
+
+  | Comp c ->
+    let universe = List.map (fun u -> resugar_universe u) c.comp_univs in
+    let result = (resugar_term c.result_typ, A.Nothing) in
+    let args = 
+      if (lid_equals c.effect_name C.effect_Lemma_lid) then
+        let rec aux l = function
+          | [] -> l
+          | (t,aq)::tl -> 
+            match (t.n) with
+            | Tm_fvar fv when S.fv_eq_lid fv C.true_lid ->
+              aux l tl
+            | Tm_meta _ (* where metadata == Meta_desuagard(Meta_smt_pat) *) ->
+              aux l tl
+            | _ -> 
+              aux ((t,aq)::l) tl
+        in
+        aux [] c.effect_args
+      else 
+        c.effect_args 
+    in
+    let args = List.map(fun (e,_) -> (resugar_term e, A.Nothing)) args in
+    let rec aux l = function 
+      | [] -> l
+      | hd::tl ->
+        match hd with 
+          | DECREASES e ->
+            let e = (resugar_term e, A.Nothing) in
+            aux (e::l) tl
+          | _ -> aux l tl
+    in
+    let decrease = aux [] c.flags in
+    mk (A.Construct(c.effect_name, decrease@args))
```

Comment:
I don't think there is a reverse. It is trying to concatenate terms from comp_typ.effect_args and comp_typ.flags.